samples/matter/common/src/matter_init.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "matter_init.h"
+
+#include "fabric_table_delegate.h"
+
+#ifdef CONFIG_CHIP_OTA_REQUESTOR
+#include "ota_util.h"
+#endif
+
+#ifdef CONFIG_CHIP_WIFI
+#include <platform/nrfconnect/wifi/NrfWiFiDriver.h>
+#endif
+
+#ifdef CONFIG_MCUMGR_TRANSPORT_BT
+#include "dfu_over_smp.h"
+#endif
+
+#include <app/clusters/network-commissioning/network-commissioning.h>
+#include <app/server/OnboardingCodesUtil.h>
+#include <credentials/examples/DeviceAttestationCredsExample.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
+
+using namespace ::chip::DeviceLayer;
+using namespace ::chip::Credentials;
+using namespace ::chip::app;
+using namespace ::chip;
+
+namespace
+{
+#ifdef CONFIG_CHIP_WIFI
+Clusters::NetworkCommissioning::Instance sWiFiCommissioningInstance(0,
+								    &(NetworkCommissioning::NrfWiFiDriver::Instance()));
+Clusters::NetworkCommissioning::Instance *sCommissioningInstance{ &sWiFiCommissioningInstance };
+#else
+Clusters::NetworkCommissioning::Instance *sCommissioningInstance = nullptr;
+#endif
+
+#if CONFIG_CHIP_FACTORY_DATA
+FactoryDataProvider<InternalFlashFactoryData> sDefaultFactoryDataProvider;
+FactoryDataProviderProxy *sFactoryDataProvider{ &sDefaultFactoryDataProvider };
+#endif
+
+CommonCaseDeviceServerInitParams sServerInitParamsDefault;
+CommonCaseDeviceServerInitParams *sServerInitParams{ &sServerInitParamsDefault };
+
+Nordic::Matter::ChipCustomInit sCustomInit{ nullptr };
+
+K_MUTEX_DEFINE(sInitMutex);
+K_CONDVAR_DEFINE(sInitCondVar);
+CHIP_ERROR sInitResult;
+bool sInitDone{ false };
+
+/* RAII utility to implement automatic cleanup and signalling. */
+struct InitGuard {
```

Comment:
Why do we need the whole InitGuard functionality? I can see that it is "working" when the DoInitChipServer method is in progress, but what can interrupt that method? As far as I remember the whole init is run sequentially until we call `PlatformMgr().StartEventLoopTask();`. Do you observe some multithread problems here, and due to that have you decided to add the Init Guard?

Reply:
Because to achieve thread safety and synchronization we initialize all Matter interfaces in CHIP thread and all application object in the application thread. The second part can only be started once the CHIP initialization is finalized. That's why the application thread must wait until the CHIP thread signals that the initialization is done.